store/store_k8s.go

```diff
@@ -0,0 +1,257 @@
+	}
+	return &k8sStore{cm}, cm, nil
+}
+
+func (k8s *k8sStore) Lock(owner string) (*StoreLock, error) {
+	if err := k8s.cm.Lock(owner); err != nil {
+		return nil, err
+	}
+	return &StoreLock{owner: owner}, nil
+}
+
+func (k8s *k8sStore) LockWithKey(owner, key string) (*StoreLock, error) {
+	if err := k8s.cm.LockWithKey(owner, key); err != nil {
+		return nil, err
+	}
+	return &StoreLock{Key: key, owner: owner, lockedWithKey: true}, nil
+}
+
+func (k8s *k8sStore) Unlock(storeLock *StoreLock) error {
+	if storeLock.lockedWithKey {
+		return k8s.cm.UnlockWithKey(storeLock.Key)
+	}
+	return k8s.cm.Unlock()
+}
+
+func (k8s *k8sStore) IsKeyLocked(key string) (bool, string, error) {
+	return k8s.cm.IsKeyLocked(key)
+}
+
+func (k8s *k8sStore) CreateKey(key string, value []byte) error {
+	sanitizedKey := GetSanitizedK8sName(key)
+	err := k8s.cm.LockWithKey(string(value), sanitizedKey)
+	if err != nil {
+		logrus.Errorf("unable to lock with key %v", key)
+		return err
+	}
+	defer func() {
+		err := k8s.cm.UnlockWithKey(sanitizedKey)
+		if err != nil {
+			logrus.Infof("unable to unlock with key %v", key)
+		}
+	}()
+
+	data, err := k8s.cm.Get()
+	if err != nil {
+		return err
+	}
+
+	if _, ok := data[key]; ok {
+		return &StoreKeyExists{
+			Key:     key,
+			Message: "Use PutKey API",
+		}
+	}
+
+	if data == nil {
+		data = make(map[string]string)
+	}
+	data[key] = string(value)
+	return k8s.patchWithRetries(data)
+}
+
+func (k8s *k8sStore) PutKey(key string, value []byte) error {
+	data, err := k8s.cm.Get()
+	if err != nil {
+		return err
+	}
+
+	data[key] = string(value)
+	return k8s.patchWithRetries(data)
+}
+
+func (k8s *k8sStore) GetKey(key string) ([]byte, error) {
+	data, err := k8s.cm.Get()
+	if err != nil {
+		return nil, err
+	}
+
+	value, ok := data[key]
+	if !ok {
+		return nil, &StoreKeyDoesNotExist{
+			Key: key,
+		}
+	}
+
+	return []byte(value), nil
+}
+
+func (k8s *k8sStore) DeleteKey(key string) error {
+	sanitizedKey := GetSanitizedK8sName(key)
+	// Let's use the sanitized key itself as the owner.
+	err := k8s.cm.LockWithKey(sanitizedKey, sanitizedKey)
+	if err != nil {
+		logrus.Errorf("unable to lock with key %v", key)
+		return err
+	}
+	defer func() {
+		err := k8s.cm.UnlockWithKey(sanitizedKey)
+		if err != nil {
+			logrus.Infof("unable to unlock with key %v", key)
+		}
+	}()
+	data, err := k8s.cm.Get()
+	if err != nil {
+		return err
+	}
+
+	if _, ok := data[key]; !ok {
+		return nil
+	}
+
+	delete(data, key)
+	return k8s.cm.Update(data)
+}
+
+func (k8s *k8sStore) EnumerateKey(key string) ([]string, error) {
+	data, err := k8s.cm.Get()
+	if err != nil {
+		return nil, err
+	}
+
+	returnKeys := make([]string, 0)
+	for k := range data {
+		if strings.HasPrefix(k, key) {
+			returnKeys = append(returnKeys, k)
+		}
+	}
+
+	return returnKeys, nil
+}
+
+func (k8s *k8sStore) patchWithRetries(data map[string]string) error {
+	f := func() (bool, error) {
+		err := k8s.cm.Patch(data)
+
+		for _, retryErr := range etcdErrorsToRetry {
```

Comment:
do we actually see etcd errors from the k8s api server? I would have thought that the k8s API server will hide etcd errors from the clients.

Reply:
I am not sure. May be @adityadani might have more context to this.

Reply:
changed